lms/djangoapps/dashboard/sysadmin.py

```diff
@@ -480,6 +485,173 @@ def post(self, request):
+
+                course_obj = CourseStructure.objects.get(course_id=course_key)  # course content tree, serialised
+
+                # deserialising and converting to OrderedDict with the blocks
+                # in the order that they appear in the course
+                course_ordered = course_obj.ordered_blocks  # OrderedDict
+
+                def get_children(parent):
+                    """
+                    Method for getting children from course structure object
+                    """
+                    children = course_ordered[parent].get('children', [])
+                    return children
+
+                # getting list of all sequentials (child for chapter)
+                sequentials = []  # The subsections defined in the course outline
+                for chapter in course_chapters:
+                    one_chapter_seqs = get_children(chapter)
+                    for seq in one_chapter_seqs:
+                        sequentials.append(seq)
+
+                # getting list of all verticals (child for sequential)
+                verticals = []  # The units defined in the course outline
+                for sequential in sequentials:
+                    one_seq_verts = get_children(sequential)
+                    for vert in one_seq_verts:
+                        verticals.append(vert)
+
+                # getting vertical (unit) names
+                unit_names = []
+                for vertical in verticals:
+                    unit_name = course_ordered[vertical].get('display_name')
+                    unit_names.append(unit_name)
+
+                header = [_('username'), _('email'), _('registration date'), _('enrolled courses'), ('last login'), 'last visit', ]
+
+                # preparing list for counting visits for course units (last raw in the table)
+                visit_count = ['', '', '', '', '', 'Visits:', ]
+
+                for name in unit_names:
+                    header.append(name)  # adding unit names to the table header
+                    visit_count.append(0)
+
+                for u in enrolled_students:  # u - User object, enrolled to current course
+                    db_query = models.StudentModule.objects.filter(
+                        course_id__exact=course_key,
+                        student_id=u.id
+                    )   # querySet StudentModule
+                    last_visit_id = db_query.all().order_by('-modified').values_list('module_state_key', flat=True)[:1]
+                    try:
+                        last_visit = course_ordered[last_visit_id[0]].get('display_name')
+                    except Exception:
+                        last_visit = 'New to course'
+
+                    visited_in_course = []
+                    for block in db_query.all().values('module_state_key'):
+                        msk = block['module_state_key']
+                        visited_in_course.append(msk)  # usageKeys for course parts where student have been
```

Comment:
visited_in_course = [b['module_state_key'] for b in db_query.all().values('module_state_key')] ?